README
NAME
    "Convert::Color::Library" - named lookup of colors from "Color::Library"

SYNOPSIS
    Directly:

     use Convert::Color::Library;

     my $red = Convert::Color::Library->new( 'red' );

     # Only use the SVG dictionary
     my $brown = Convert::Color::Library->new( 'svg/brown' );

     # Use either HTML or SVG dictionary
     my $pink = Convert::Color::Library->new( 'html,svg/pink' );

    Via Convert::Color:

     use Convert::Color;

     my $cyan = Convert::Color->new( 'lib:cyan' );

     my $darkcyan = Convert::Color->new( 'lib:windows/darkcyan' );

DESCRIPTION
    This subclass of Convert::Color::RGB8 provides lookup of color names
    using Robert Krimen's Color::Library module. It therefore provides
    convenient access to named colours in many dictionaries, such as SVG,
    X11 and HTML.

CONSTRUCTOR
  $color = Convert::Color::Library->new( $name )
    Returns a new object to represent the named color.

    If the name is of the form

     dicts/name

    Then "dicts" is parsed as a comma-separated list of dictionary names to
    pass to "Color::Library".

TODO
    *   Consider an API for getting the list of dictionary names and colour
        names. That said, it's easy enough to do directly to
        "Color::Library", so maybe not needed.

SEE ALSO
    *   Convert::Color - color space conversions

    *   Color::Library - An easy-to-use and comprehensive named-color
        library

AUTHOR
    Paul Evans <leonerd@leonerd.org.uk>



Changes
Revision history for Convert-Color-Library

0.04    2014/11/13 19:15:06
        [CHANGES]
         * Expose Convert::Color::HTML as a toplevel colorspace; give it
           HTML-specific abilities with #RRGGBB notation
         * General distribution updates

0.03    CHANGES:
         * Updated for Convert::Color 0.06

0.02    CHANGES:
         * Added 'use warnings'
         * Various updates to keep CPANTS happy

0.01    First version, released on an unsuspecting world.



lib/Convert/Color/HTML.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2014 -- leonerd@leonerd.org.uk

package Convert::Color::HTML;

use strict;
use warnings;
use base qw( Convert::Color::Library );

# Ensure that Color::Library::Dictionary::HTML is loaded
use Color::Library;

__PACKAGE__->register_color_space( 'html' );

our $VERSION = '0.04';

=head1 NAME

C<Convert::Color::HTML> - color conversion using C<Color::Library::Dictionary::HTML>

=head1 SYNOPSIS

Directly:

 use Convert::Color::HTML;

 my $red = Convert::Color::HTML->new( 'red' );

 my $blue = Convert::Color::HTML->new( '#0000FF' );

Via L<Convert::Color>:

 use Convert::Color;

 my $cyan = Convert::Color->new( 'html:cyan' );

Conversion from RGB:

 my $green = Convert::Color::RGB->( 0, 1.0, 0 )->as_html;
 say "HTML colour name is " . $green->name;

=head1 DESCRIPTION

This subclass of L<Convert::Color::Library> provides a shortcut to performing
library lookups specifically within the C<HTML> dictionary. Additionally it
will parse C<#RRGGBB> color specifications.

=cut

=head1 CONSTRUCTOR

=cut

=head2 $color = Convert::Color::HTML->new( $name )

Returns a new object to represent the named color from the C<HTML> dictionary
or plain RGB triplet.

=cut

sub new
{
   my $class = shift;
   my ( $name ) = @_;

   if( $name =~ m/^#([0-9a-f]{6})$/i ) {
      my $self = $class->Convert::Color::RGB8::new( $1 );
      $self->[3] = uc "#$1";
      return $self;
   }

   return $class->SUPER::new( "html/$name" );
}

=head1 METHODS

=cut

=head2 $name = $color->name

Returns the name of the color instance; either the C<HTML> dictionary name, or
the plain RGB triplet notation, as was passed to the constructor.

=cut

# Don't register this as a palette space, because we can always represent any
# RGB8 colour. Instead, provide the methods directly

my %palette;

sub new_from_rgb8
{
   my $class = shift;
   my ( $rgb8 ) = @_;

   my $hex = $rgb8->hex;

   unless( keys %palette ) {
      %palette = map {
         my $color = $class->new( $_ );
         $color->hex => $color 
      } Color::Library::Dictionary::HTML->color_names;
   }

   return $palette{$hex} || $class->new( "#$hex" );
}

sub new_rgb
{
   my $class = shift;
   return $class->new_from_rgb8( Convert::Color::RGB->new( @_ )->as_rgb8 );
}

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;


lib/Convert/Color/Library.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2009-2014 -- leonerd@leonerd.org.uk

package Convert::Color::Library;

use strict;
use warnings;
use base qw( Convert::Color::RGB8 );

__PACKAGE__->register_color_space( 'lib' );

use Carp;

our $VERSION = '0.04';

=head1 NAME

C<Convert::Color::Library> - named lookup of colors from C<Color::Library>

=head1 SYNOPSIS

Directly:

 use Convert::Color::Library;

 my $red = Convert::Color::Library->new( 'red' );

 # Only use the SVG dictionary
 my $brown = Convert::Color::Library->new( 'svg/brown' );

 # Use either HTML or SVG dictionary
 my $pink = Convert::Color::Library->new( 'html,svg/pink' );

Via L<Convert::Color>:

 use Convert::Color;

 my $cyan = Convert::Color->new( 'lib:cyan' );

 my $darkcyan = Convert::Color->new( 'lib:windows/darkcyan' );

=head1 DESCRIPTION

This subclass of L<Convert::Color::RGB8> provides lookup of color names
using Robert Krimen's L<Color::Library> module. It therefore provides
convenient access to named colours in many dictionaries, such as SVG, X11 and
HTML.

=cut

=head1 CONSTRUCTOR

=cut

=head2 $color = Convert::Color::Library->new( $name )

Returns a new object to represent the named color.

If the name is of the form

 dicts/name

Then C<dicts> is parsed as a comma-separated list of dictionary names to pass
to C<Color::Library>.

=cut

sub new
{
   my $class = shift;

   require Color::Library;

   @_ == 1 or
      croak "usage: Convert::Color::Library->new( NAME )";

   my ( $name ) = @_;

   my $color;

   if( $name =~ m{^(.*)/(.*)$} ) {
      ( my $dicts, $name ) = ( $1, $2 );

      $color = Color::Library->color( [ split m/,/, $dicts ], $name );
   }
   else {
      $color = Color::Library->color( $name );
   }

   defined $color or croak "No such library color named '$name'";
   my $self = $class->SUPER::new( $color->rgb );

   $self->[3] = $color->name;
   $self->[4] = $color->dictionary->name;

   return $self;
}

=head1 METHODS

=cut

=head2 $name = $color->name

=head2 $dict = $color->dict

Returns the name of the color within its dictionary, and the name of the
dictionary itself.

=cut

sub name { shift->[3] }
sub dict { shift->[4] }

=head1 TODO

=over 4

=item *

Consider an API for getting the list of dictionary names and colour names.
That said, it's easy enough to do directly to C<Color::Library>, so maybe not
needed.

=item *

Expose other dictionaries (SVG? Windows?) as named spaces, like the HTML one.

=back

=head1 SEE ALSO

=over 4

=item *

L<Convert::Color> - color space conversions

=item *

L<Color::Library> - An easy-to-use and comprehensive named-color library

=item *

L<Convert::Color::HTML> - color conversion using C<Color::Library::Dictionary::HTML>

=back

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;
